keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,770 @@
+   // such as `quiesce` are reserved for Kubernetes-defined “well-known” names.
+   // Names with a prefix such as `example.com/label-style` are custom names.
+-   // This refers to a ContainerNotifierAction and a Probe.
++   // This refers to a ContainerNotifierAction and a Probe.
++   // Kubelet executes the command defined in ContainerNotifierAction, and then
++   // calls the command defined in Probe to verify whether the action has  
++   // resulted in a desired state.
++   // For example, if the ContainerNotifierAction has a command to quiesce the application,
++   // then the Probe has a command to verify that the application is indeed
++   // being quiesced. Kubelet will run the command to do a quiesce and wait for the quiescent
++   // probe to return success.
+   Name string
+   Action *ContainerNotifierAction
++   // Add a Probe in Phase 2. Probe defined in the core API here will be used.
++   Probe *Probe
+}
+```
+
+### Kubelet Impact in Phase 2 and Beyond
+
+When moving the logic from the container notifier controller to Kubelet, there will be impact on Kubelet.
+
+Kubelet watches new Notification resources cluster-wide.
+
+Kubelet must execute Notification command/signal against containers. CRI changes are required to support signals.
+
+Kubelet must update Notification status (possible QPS issues trying to update for 100+ of pods).
+
+Kubelet must have some method of measuring the success/failure of the Notification.
+* For exec, it depends on the return value of the ExecInContainer method.
+* For signals, it depends on the new CRI changes.
+
+Kubelet will not retry if the ContainerNotifier call fails. It is up to the external controller who requested the ContainerNotifier to send another request to retry.
+
+#### CRI Changes
+
+To support signals directly, we need to make changes in CRI. We can add a new "Signal" Runtime interface with an input parameter "signal string" and handle a specified number of signals, starting with "SIGHUP". For example, if the input parameter "signal" is "SIGHUP", it is a signal to kill the container.
+
+``` 
+// Runtime is the interface to execute the commands and provide the streams.
+type Runtime interface {
+        Exec(containerID string, cmd []string, in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize) error
++        Signal(containerID string, signal string, in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize) error
+
+        Attach(containerID string, in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize) error
+        PortForward(podSandboxID string, port int32, stream io.ReadWriteCloser) error
+}
+```
+
+To support "SIGHUP" directly, we can either convert it to the command "kill -SIGHUP" and send it the same way as other commands, or we can translate that into a call to the docker client method `ContainerKill`.
```

Comment:
you can't assume a container has a `kill` binary to execute

Reply:
Removed wording using the command.